internal/storage/google_cloud_storage_test.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+	"testing"
+
+	"cloud.google.com/go/storage"
+)
+
+func testGoogleCloudStorageClient(tb testing.TB) *storage.Client {
+	tb.Helper()
+
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		tb.Fatal(err)
+	}
+	return client
+}
+
+func testName(tb testing.TB) string {
+	tb.Helper()
+
+	var b [512]byte
+	if _, err := rand.Read(b[:]); err != nil {
+		tb.Fatalf("failed to generate random: %v", err)
+	}
+	digest := fmt.Sprintf("%x", sha256.Sum256(b[:]))
+	return digest[:32]
+}
+
+func testGoogleCloudStorageBucket(tb testing.TB) string {
+	tb.Helper()
+
+	if testing.Short() {
+		tb.Skipf("🚧 Skipping Google Cloud Storage tests (short!")
+	}
+
+	if skip, _ := strconv.ParseBool(os.Getenv("SKIP_GOOGLE_CLOUD_STORAGE_TESTS")); skip {
+		tb.Skipf("🚧 Skipping Google Cloud Storage tests (SKIP_GOOGLE_CLOUD_STORAGE_TESTS is set)!")
+	}
+
+	bucketID := os.Getenv("GOOGLE_CLOUD_BUCKET")
+	if bucketID == "" {
+		tb.Fatal("missing GOOGLE_CLOUD_BUCKET!")
+	}
+
+	return bucketID
+}
+
+func testGoogleCloudStorageObject(tb testing.TB, r io.Reader) string {
+	tb.Helper()
+
+	ctx := context.Background()
+	client := testGoogleCloudStorageClient(tb)
+	bucket := testGoogleCloudStorageBucket(tb)
+	name := testName(tb)
+
+	// Create the object.
+	w := client.Bucket(bucket).Object(name).NewWriter(ctx)
+	if _, err := io.Copy(w, r); err != nil {
+		tb.Fatalf("failed to create object: %v", err)
+	}
+	if err := w.Close(); err != nil {
+		tb.Fatalf("failed to close writer: %v", err)
+	}
+
+	// Schedule cleanup.
+	tb.Cleanup(func() {
+		if err := client.Bucket(bucket).Object(name).Delete(ctx); err != nil && err != storage.ErrObjectNotExist {
+			tb.Fatalf("failed cleaning up %s: %v", name, err)
+		}
+	})
+
+	return name
+}
+
+func TestGoogleCloudStorage_CreateObject(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+	client := testGoogleCloudStorageClient(t)
+	bucket := testGoogleCloudStorageBucket(t)
+	object := testGoogleCloudStorageObject(t, strings.NewReader("contents"))
+
+	cases := []struct {
+		name     string
+		bucket   string
+		object   string
+		contents []byte
+		err      bool
+	}{
+		{
+			name:     "default",
+			bucket:   bucket,
+			object:   "myfile",
```

Comment:
Since all prow jobs use the same bucket passed in as `GOOGLE_CLOUD_BUCKET `, if there are multiple presubmit prow jobs running at the same time, there will be a chance of collision on this object, can we utilize the same random hash layout to get the gcs bucket better organized?

Reply:
Yup - you're right. We can use the `testName` helper to get uniqueness. Let me know when you want me to update and re-test.

Reply:
We can start test once https://github.com/GoogleCloudPlatform/oss-test-infra/pull/403 is merged, please update the test

Reply:
Done